File: Core/src/main/java/net/winternetwork/bedwars/core/game/stage/ToStartStage.java
package net.winternetwork.bedwars.core.game.stage;

import net.winternetwork.bedwars.api.game.flag.Flag;
import net.winternetwork.bedwars.api.game.stage.Stage;
import net.winternetwork.bedwars.api.score.ReplaceableList;
import net.winternetwork.bedwars.core.game.GameSettings;
import net.winternetwork.bedwars.core.game.flag.Flags;
import net.winternetwork.bedwars.core.game.map.GameMap;
import net.winternetwork.bedwars.core.game.map.MapManager;
import org.bukkit.entity.Player;

import java.util.Arrays;
import java.util.List;

public class ToStartStage extends Stage {


    public ToStartStage() {
        super("Começando.", 20);
    }

    private boolean started = false;

    @Override
    public List<Flag> getFlags() {
        return Arrays.asList(
                Flags.NO_PVP,
                Flags.NO_BUILD
        );
    }

    @Override
    public ReplaceableList getScoreboard() {
        return new ReplaceableList(
                Arrays.asList(
                        "&1",
                        "&fMapa: &a<mapa>",
                        "&fComeçando em: &a<next>",
                        "&2",
                        "&fJogadores: &a<online>/<max>",
                        "&3",
                        "&ewww.baianowars.kt"
                ),
                (s, player) -> s
                        .replaceAll("<mapa>", "Void")
                        .replaceAll("<online>", getOnlinePlayers() + "")
                        .replaceAll("<max>", "15")
                        .replaceAll("<next>", String.format("%ss", getTimeLeft()))
        );
    }

    @Override
    public void onStageJoin() {
    }

    @Override
    public void onSecondPassed() {

        if (getTimeLeft() <= 0) {
            StageManager.getInstance().callNextStage();
        }

        if (getTimeLeft() <= 5) {
            broadcast(
                    String.format("Começando o jogo em %ss.", getTimeLeft())
            );
        }
    }

    @Override
    public void onTimeLeft() {
        if (getOnlinePlayers() < GameSettings.PLAYERS_TO_START) {
            broadcast("§eAinda não tem jogadores suficientes para começar.");

            StageManager.getInstance().callPreviousStage();
            return;
        }

        started = true;
        StageManager.getInstance().callNextStage();
    }

    @Override
    public void onStageExit() {
        if (!started) return;
        broadcast("Começou!");

        GameMap map = MapManager.getInstance().getAll().get(0);

        System.out.println(map.toString());

        for (int i = 0; i < getAllPlayers().size(); i++) {
            Player player = getAllPlayers().get(i);

            player.teleport(map.getSpawnLocations()[i]);
        }
    }
}

File: Game/src/main/kotlin/net/winternetwork/bedwars/modules/stage/object/ToStartStage.kt
package net.winternetwork.bedwars.modules.stage.`object`

import net.winternetwork.bedwars.api.game.stage.Stage
import net.winternetwork.bedwars.api.score.ReplaceableList
import net.winternetwork.bedwars.game.settings.GameSettings
import net.winternetwork.bedwars.modules.maps.MapModule
import net.winternetwork.bedwars.modules.stage.StageModule
import net.winternetwork.bedwars.modules.stage.flag.Flags

class ToStartStage(
    stageModule: StageModule,
    mapModule: MapModule
) : Stage("Starting", 20) {

    private val stageManager = stageModule.stageManager

    private val mapManager = mapModule.manager

    private var started = false

    override val flags = listOf(
        Flags.NO_PVP_FLAG,
        Flags.NO_BREAK_FLAG,
        Flags.NO_BUILD_FLAG,
        Flags.NO_GENERATOR_FLAG
    )

    override val scoreboard = ReplaceableList(
        listOf(
            "&1",
            "&fMapa: &a<mapa>",
            "&fComeçando em: &a<next>",
            "&2",
            "&fJogadores: &a<online>/<max>",
            "&3",
            "&ewww.baianowars.kt"
        )
    ) {
        mapOf(

            "<mapa>" to "Gadolandia",
            "<online>" to "$onlinePlayers",
            "<max>" to "${GameSettings.MAX_PLAYERS}",
            "<next>" to "${timeLeft}s"
        )
    }

    override fun onSecondPassed() {
        when {
            timeLeft <= 0 -> {
                stageManager!!.next()
                return
            }
            timeLeft <= 5 -> broadcast("§eComeçando o jogo em ${timeLeft}s")
        }
    }

    override fun onTimeLeft() {
        if (onlinePlayers < GameSettings.PLAYERS_TO_START) {
            broadcast(
                "§eAinda não tem jogadores suficientes para começar."
            )
            stageManager!!.previous()
            return
        }

        started = true
        stageManager!!.next()
    }

    override fun onStageExit() {
        if (!started) return
        broadcast("§2Começou!")

        val map = mapManager.all().first()

        allPlayers.forEachIndexed { i, p ->
            p.teleport(map.locations[i])
        }
    }

}